zaptest/observer/observer.go

```diff
@@ -80,6 +80,39 @@ func (o *ObservedLogs) AllUntimed() []LoggedEntry {
 	return ret
 }
 
+// FilterMessage filters entries to those that have the specified message.
+func (o *ObservedLogs) FilterMessage(msg string) *ObservedLogs {
+	return o.filter(func(e LoggedEntry) bool {
+		return e.Message == msg
+	})
+}
+
+// FilterField filters entries to those that have the specified field.
+func (o *ObservedLogs) FilterField(field zapcore.Field) *ObservedLogs {
+	return o.filter(func(e LoggedEntry) bool {
+		for _, ctxField := range e.Context {
+			if ctxField.Key != field.Key {
+				continue
+			}
+			return ctxField == field
```

Comment:
Sorry, I should be a bit clearer. It's entirely possible to have multiple fields with the same key. Consider

```
logger.With(zap.Int("foo", 1), zap.Namespace("component"), zap.Int("foo", 2))
```

and a filter on `zap.Int("foo", 2)`.

Given that, I don't think we want to explicitly compare keys; instead, I'd expect something more like

```
for _, f := range e.Context {
  if f == field {
    return true
  }
}
return false
```